### PR TITLE
Issue 4942: [2D MVP] Milestone 3: 2D Vector drawing

### LINKED ISSUE #4942: [2D MVP] Milestone 3: 2D Vector drawing
### Description
For calibrating and for future features we need a way to add UI elements to the drawing that respond to panning/zooming. In this issue we provide that ui layer and vector drawing.

### Goals
- [ ] As a user I want to be able to draw a vector in the 2d viewer and that it responds appropriately to pan/zoom

### Tasks
- [ ] Create a component ``<2DViewerLayer {..props}/>`` that sits on top of the 2d viewer
- [ ]  props of 2DViewerLayer  are children (svg ui elements), **viewbox** (a rect that will represent the pan/zoom position of the drawing

### Related Resources
[MockUp](https://www.figma.com/file/uTWnaYZ8IyoTn2i2xKtdBo/2D---3D?type=design&node-id=7-35530&mode=design&t=013L82bg5wfSjJMa-0)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #4942 

#### Description
Create a layer to draw arrow on top of the 2d viewer

#### Test cases
Play around with the arrow, check it creates correctly, the app doesn't crush, and the coordinates are correct

### CODE REVIEW

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/components/viewer/drawingViewer/drawingViewer.component.tsx [hunk 1]
```diff
@@ -49,19 +54,45 @@ export const DrawingViewer = () => {
...
+	useEffect(() => {
+		if (!zoomHandler) return;
+		zoomHandler.on(Events.transform, () => {
+			const transform = zoomHandler.getTransform();
+			const { scale } = transform;
+			const cantZoomOut = zoomHandler.getMinZoom() >= scale;
+			const cantZoomIn = zoomHandler.getMaxZoom() <= scale;
 			setIsMinZoom(cantZoomOut);
 			setIsMaxZoom(cantZoomIn);
+			setViewBox({ ...transform, ...zoomHandler.getOriginalSize() });
```
[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/src/v5/ui/components/viewer/drawingViewer/drawingViewer.component.tsx:75]: also.. why do you need the original size?

File: frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+	// This returns the offset of the cursor from the top-left corner
+	const getCursorOffset = (e) => {
+		const rect = e.target.getBoundingClientRect();
+		const offsetX = e.clientX - rect.left;
+		const offsetY = e.clientY - rect.top;
+		return [offsetX, offsetY].map((point) => point / viewBox.scale) as Offset;
+	};
+
+	const handleMouseDown = () => previousViewBox.current = viewBox;
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx:52]: It takes a couple of clicks before starting to draw
drawing arrow

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/components/viewer/drawingViewer/drawingViewer.component.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/drawingViewer/drawingViewer.component.tsx:75]: That's for the drawing layer to know how big it should be.
The transform object only contains info about the zoom level and the offsets, but not about the size

[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/drawingViewer/drawingViewer.component.tsx:75]: The transform object only contains the offset and the zoom level, which are not enough for the drawingContainer to know where the image ends, and so what are the boundaries to draw the arrow (or hovering circle), to know that, it needs the original image sizes

File: frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	active: boolean,
+	onChange?: (arrow: { start: Offset, end: Offset }) => void;
+};
+export const ViewerLayer2D = ({ viewBox, active, onChange }: ViewerLayer2DProps) => {
+	const [offsetStart, setOffsetStart] = useState<Offset>(null);
+	const [offsetEnd, setOffsetEnd] = useState<Offset>(null);
+	const previousViewBox = useRef<ViewBoxType>(null);
+	const [mousePosition, setMousePosition] = useState<Offset>(null);
+
+	const layerStyle: CSSProperties = {
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx:37]: This is current code. The panzoom object modifies directly the image via css `transform: matrix`. In order to adjust the other layers to overlap the image, they also need to adjust their transformation to it

File: frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/svgCircle/svgCircle.component.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import { Circle, Svg } from './svgCircle.styles';
+
+type SvgCircleProps = { coords: Offset, scale: number };
+export const SvgCircle = ({ coords, scale }: SvgCircleProps) => {
+	const measures = mapValues({
+		strokeWidth: 1,
+		radius: 3,
+	}, (val) => val / scale);
+	return (
+		<Svg
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/svgCircle/svgCircle.component.tsx:29]: I couldn't see noticeable improvements

File: frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
+/**
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx:1]: Pan and zoom using the mouse no longer works

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by The-Daniel — APPROVED** [review]

**Pull request merged**